### ISSUE TITLE
Ignore orphaned pods

### ISSUE THREAD

**Issue by irasnyd** [FairwindsOps/polaris#327]
### Installation Process
 

EKS cluster, version 1.15, using helm.

Command line:
```bash
$ helm upgrade --install polaris fairwinds-stable/polaris --version 1.0.2 --namespace kube-system -f polaris-values.yaml
```

`polaris-values.yaml` contents:
```yaml
# Override the tag provided in the chart as the unconstrained "1.0"
# to be the more constrained "1.0.3" (a specific release).
image:
  tag: "1.0.3"

# Enable the dashboard over HTTP for internal users
dashboard:
  enable: true
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx-ingress-private
    hosts:
      - "polaris.internal.domain"
```

### Polaris Version

Version 1.0.3 Docker Image

```
$ k -n kube-system get deploy polaris-dashboard -o wide
NAME                READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS   IMAGES                            SELECTOR
polaris-dashboard   1/1     1            1           15d   dashboard    quay.io/fairwinds/polaris:1.0.3   app=polaris,app.kubernetes.io/instance=polaris,app.kubernetes.io/name=polaris,component=dashboard

```

### Expected Behavior

The dashboard should be working.

### Actual Behavior

Webpage with contents:

```
Error fetching Kubernetes resources
```

Logs from container:
```
$ k -n kube-system logs polaris-dashboard-dcb6c8b9c-9bpkw 
time="2020-06-04T17:36:42Z" level=info msg="Starting Polaris dashboard server on port 8080"
time="2020-06-04T17:37:04Z" level=error msg="Cache missed ReplicaSet/prod/telbot-gunicorn-7844c9db9d again"
time="2020-06-04T17:37:04Z" level=error msg="Error loading controllers from pods: Could not retrieve parent object"
time="2020-06-04T17:37:04Z" level=error msg="Error fetching Kubernetes resources Could not retrieve parent object"
```

CURL output of same request:
```
$ curl -v -L 'http://polaris.internal.domain/'
* About to connect() to polaris.internal.domain port 80 (#0)
*   Trying 10.100.25.113...
* Connected to polaris.internal.domain (10.100.25.113) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: polaris.internal.domain
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Date: Thu, 04 Jun 2020 17:44:50 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Length: 36
< Connection: keep-alive
< Server: nginx/1.17.8
< X-Content-Type-Options: nosniff
< 
Error fetching Kubernetes resources
* Connection #0 to host polaris.internal.domain left intact
```

### Steps to Reproduce

1. Install with helm.
2. Open dashboard web interface.
3. Get `Error fetching Kubernetes resources` error message

### Additional information

I'm not sure what further information to provide to help diagnose this issue.

**Comment by baderbuddy** [comment]
Hi there, Polaris seems to be having issues pulling some metadata from your Kubernetes cluster. Is there anything special about that ReplicaSet that it's erroring out on?

**Comment by irasnyd** [comment]
There should be nothing special about the named ReplicaSet.

I do notice something very strage, however, the ReplicaSet mentioned in the error message doesn't exist! WTF?

```
$ k -n kube-system logs polaris-dashboard-dcb6c8b9c-jclfh 
time="2020-06-04T18:01:58Z" level=info msg="Starting Polaris dashboard server on port 8080"
time="2020-06-04T18:02:23Z" level=error msg="Cache missed ReplicaSet/prod/telbot-gunicorn-7844c9db9d again"
time="2020-06-04T18:02:23Z" level=error msg="Error loading controllers from pods: Could not retrieve parent object"
time="2020-06-04T18:02:23Z" level=error msg="Error fetching Kubernetes resources Could not retrieve parent object"
$ k -n prod get rs --sort-by=.metadata.creationTimestamp | grep telbot-gunicorn 
telbot-gunicorn-568bbdc45f                          0         0         0       2d21h
telbot-gunicorn-86668c896                           0         0         0       2d21h
telbot-gunicorn-656f5d65d                           0         0         0       2d19h
telbot-gunicorn-76dc6779bc                          0         0         0       2d19h
telbot-gunicorn-5656d7586d                          0         0         0       2d19h
telbot-gunicorn-59877647c4                          0         0         0       2d18h
telbot-gunicorn-6d5dd64b87                          0         0         0       2d18h
telbot-gunicorn-7f56bbccff                          0         0         0       2d17h
telbot-gunicorn-5549dc776d                          0         0         0       47h
telbot-gunicorn-85f7bf9f8c                          0         0         0       47h
telbot-gunicorn-69779cb9c6                          2         2         2       21h
$ k -n prod get rs --sort-by=.metadata.creationTimestamp | grep telbot-gunicorn | grep 7844
```

WTF? The ReplicaSet `telbot-gunicorn-7844c9db9d` just doesn't exist. Let's check all namespaces:
```
$ k get rs --all-namespaces | grep telbot-gunicorn
prod            telbot-gunicorn-5549dc776d                             0         0         0       47h
prod            telbot-gunicorn-5656d7586d                             0         0         0       2d19h
prod            telbot-gunicorn-568bbdc45f                             0         0         0       2d21h
prod            telbot-gunicorn-59877647c4                             0         0         0       2d18h
prod            telbot-gunicorn-656f5d65d                              0         0         0       2d19h
prod            telbot-gunicorn-69779cb9c6                             2         2         2       21h
prod            telbot-gunicorn-6d5dd64b87                             0         0         0       2d18h
prod            telbot-gunicorn-76dc6779bc                             0         0         0       2d19h
prod            telbot-gunicorn-7f56bbccff                             0         0         0       2d17h
prod            telbot-gunicorn-85f7bf9f8c                             0         0         0       47h
prod            telbot-gunicorn-86668c896                              0         0         0       2d21h
$ k get rs --all-namespaces | grep telbot-gunicorn-7844c9db9d
```

So this is really weird. Polaris has found a ReplicaSet which doesn't exist. I *did* completely delete and re-create the Polaris dashboard deployment, just in case something was cached improperly.

What can I do to help you find more information to troubleshoot this?

**Comment by rbren** [comment]
Ohh interesting. I think you have a Pod that's _owned_ by that ReplicaSet sitting around somewhere. I'm not sure if that's a k8s bug, or a case we should have been able to handle

**Comment by rbren** [comment]
I think this command will show the offending pod:

`k get po -n prod -oyaml | yq r - 'items[*].metadata.ownerReferences[*].name'`

**Comment by irasnyd** [comment]
@rbren that did it. I was able to find a Pod which was stuck in the Terminating state for >24h. A normal `kubectl delete po podname --now` didn't work, but a `k delete po podname --force --grace-period=0` did it.

And now Polaris dashboard loads perfectly! Thanks for the help.

**Comment by rbren** [comment]
I'm leave this open and change the title. We should be more resilient to cases like this

**Comment by samuelb** [comment]
I stumbled over a similar case. We had a pod stuck in termination state for days where the related ReplicaSet was gone. With Configuring a custom controller check, polaris stopped working. The Polaris dashboard showed "Error running audit", the logs of Polaris
```
WARN[0014] An error occured validating controller:error parsing JSON bytes: unexpected end of JSON input
ERRO[0014] Error getting audit data: error parsing JSON bytes: unexpected end of JSON input
```

This problem error was only shown when having a custom check for the Controller configured. Removing the custom check or the orphaned pod resolves the issue.

I tracked down the issue to the function applyControllerSchemaChecks() when it calls [check.CheckController(controller.OriginalObjectJSON)](https://github.com/FairwindsOps/polaris/blob/master/pkg/validator/schema.go#L157).

**Repro:**

Configure a custom controller check like:
```
checks:
  myControllerCheck: warning

customChecks:
  myControllerCheck:
    successMessage: This is a test
    failureMessage: This is a test
    target: Controller
    schema:
      '$schema': http://json-schema.org/draft-07/schema
      type: object
      required:
        - spec
```

Deploy something which won't terminate immediately when deleting the deployment:
```
cat <<EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: never-ending
  name: never-ending
spec:
  replicas: 1
  selector:
    matchLabels:
      app: never-ending
  template:
    metadata:
      labels:
        app: never-ending
    spec:
      containers:
      - image: busybox
        name: busybox
        command:
          - sleep
          - infinity
      terminationGracePeriodSeconds: 3600
EOF
```
And remove the deployment with `kubectl delete deployment never-ending`. This leaves a pod in termination state for one hour. Run polaris to see the errors.

**Comment by rbren** [comment]
Can you share which version of Polaris you're running?

I applied the sample config you shared for `never-ending` (thanks for that!) and ran `go run main.go audit --output-file=out.json` on the master branch. The command finishes successfully, and the output file shows a result for `ReplicaSet/never-ending`

Am I doing something differently?

**Comment by samuelb** [comment]
The version I used is 1.1.0 and the yesterdays master branch.
Did you also had the customCheck I mentioned configured?

**Comment by rbren** [comment]
Aha! Thanks, forgot to enable a controller-level check 🤦 

I think [this fix](https://github.com/FairwindsOps/polaris/pull/371/files) will be a little more robust - we'll still run the pod-level checks on that resource (which I think is important, since the Pod can still be accessed)

**Comment by samuelb** [comment]
I agree. I'll close my PR then.

**Comment by rbren** [comment]
👍 merged and released in `1.1.1` - thanks all for the patience here!
